be/src/exec/pipeline/hdfs_chunk_source.h

```diff
@@ -0,0 +1,134 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <utility>
+
+#include "exec/pipeline/chunk_source.h"
+#include "exec/vectorized/hdfs_scan_node.h"
+#include "exec/vectorized/hdfs_scanner.h"
+#include "exec/workgroup/work_group_fwd.h"
+#include "exprs/expr.h"
+#include "exprs/expr_context.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks {
+
+class SlotDescriptor;
+
+namespace vectorized {
+class RuntimeFilterProbeCollector;
+class HdfsScanner;
+} // namespace vectorized
+
+namespace pipeline {
+
+class ScanOperator;
+class HdfsChunkSource final : public ChunkSource {
+public:
+    HdfsChunkSource(MorselPtr&& morsel, ScanOperator* op, vectorized::HdfsScanNode* scan_node);
+
+    ~HdfsChunkSource() override;
+
+    Status prepare(RuntimeState* state) override;
+
+    void close(RuntimeState* state) override;
+
+    bool has_next_chunk() const override;
+
+    bool has_output() const override;
+
+    virtual size_t get_buffer_size() const override;
+
+    StatusOr<vectorized::ChunkPtr> get_next_chunk_from_buffer() override;
+
+    Status buffer_next_batch_chunks_blocking(size_t chunk_size, bool& can_finish) override;
+    Status buffer_next_batch_chunks_blocking_for_workgroup(size_t chunk_size, bool& can_finish, size_t* num_read_chunks,
+                                                           int worker_id, workgroup::WorkGroupPtr running_wg) override;
+
+private:
+    // Yield scan io task when maximum time in nano-seconds has spent in current execution round.
+    static constexpr int64_t YIELD_MAX_TIME_SPENT = 100'000'000L;
+    // Yield scan io task when maximum time in nano-seconds has spent in current execution round,
+    // if it runs in the worker thread owned by other workgroup, which has running drivers.
+    static constexpr int64_t YIELD_PREEMPT_MAX_TIME_SPENT = 20'000'000L;
```

Comment:
Maybe move this two constants to the base class `ChunkSource`? However, I'm not sure which is better, since maybe different chunk sources have different yield time.

Reply:
That is what I concern, scheduling strategies maybe are different.